transformer_engine/common/fused_attn/fused_attn.cpp

```diff
@@ -91,91 +88,99 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
           && (bias_type == NVTE_Bias_Type::NVTE_NO_BIAS)
           && (attn_mask_type == NVTE_Mask_Type::NVTE_PADDING_MASK)
-          && ((qkv_layout == NVTE_QKV_Layout::NVTE_QKV_INTERLEAVED)
-              || (qkv_layout == NVTE_QKV_Layout::NVTE_T3HD))) {
-#if (CUDNN_VERSION >= 8900)
-    backend = NVTE_Fused_Attn_Backend::NVTE_FP8;
-#else
-    backend = NVTE_Fused_Attn_Backend::NVTE_No_Backend;
-    std::cout << "Warning: FP8 fused attention is supported by cuDNN 8.9.0+."
-           " Please upgrade your cuDNN version if possible." << std::endl;
-#endif
+          && (qkv_layout == NVTE_QKV_Layout::NVTE_T3HD)) {
+    if (cudnn_runtime_version >= 8900) {
+      backend = NVTE_Fused_Attn_Backend::NVTE_FP8;
+    } else {
+      backend = NVTE_Fused_Attn_Backend::NVTE_No_Backend;
+      std::cout << "Warning: FP8 fused attention is supported by cuDNN 8.9.0+."
+          " Please upgrade your cuDNN version if possible." << std::endl;
+    }
   } else if ((q_dtype == NVTEDType::kNVTEFloat16) || (q_dtype == NVTEDType::kNVTEBFloat16)) {
     bool flag_m512 = false;
     bool flag_arb = false;
     if ((sm_arch_ == 80 || sm_arch_ == 90)
+            && (max_seqlen_q <= 512)
+            && (max_seqlen_kv <= 512)
             && (head_dim == 64)
+            && (num_attn_heads == num_gqa_groups)
             && ((bias_type == NVTE_Bias_Type::NVTE_NO_BIAS)
                 || (bias_type == NVTE_Bias_Type::NVTE_POST_SCALE_BIAS))
             && ((attn_mask_type == NVTE_Mask_Type::NVTE_CAUSAL_MASK)
-                || (attn_mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK)
                 || (attn_mask_type == NVTE_Mask_Type::NVTE_PADDING_MASK)
+                || (attn_mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK
+                    && max_seqlen_q == max_seqlen_kv)
                 || (attn_mask_type == NVTE_Mask_Type::NVTE_NO_MASK))
-            && ((qkv_layout == NVTE_QKV_Layout::NVTE_QKV_INTERLEAVED)
-                || (qkv_layout == NVTE_QKV_Layout::NVTE_KV_INTERLEAVED)
-                || (qkv_layout == NVTE_QKV_Layout::NVTE_SB3HD)
+            && ((qkv_layout == NVTE_QKV_Layout::NVTE_SB3HD)
                 || (qkv_layout == NVTE_QKV_Layout::NVTE_SBHD_SB2HD)
                 || (qkv_layout == NVTE_QKV_Layout::NVTE_BS3HD)
                 || (qkv_layout == NVTE_QKV_Layout::NVTE_BSHD_BS2HD)
                 || (qkv_layout == NVTE_QKV_Layout::NVTE_BSHD_BSHD_BSHD))) {
       flag_m512 = true;
     }
-    if (
-#if (CUDNN_VERSION >= 8903)
-        (sm_arch_ >= 80)
-#else
-        (sm_arch_ == 80 || sm_arch_ == 90)
-#endif
-            && (max_seqlen_q == max_seqlen_kv)
-            && ((head_dim == 64) || (head_dim == 128))
-            && (bias_type == NVTE_Bias_Type::NVTE_NO_BIAS)
+    if (((cudnn_runtime_version >= 8903 && sm_arch_ >= 80)
+                || (cudnn_runtime_version < 8903 && (sm_arch_ == 80 || sm_arch_ == 90)))
+            && (max_seqlen_q % 64 == 0)
+            && (max_seqlen_kv % 64 == 0)
```

Comment:
I remember the seqlen condition is `%32 == 0` after 8.9.3.

Reply:
It's supposed to, but there's a bug, so we're still pending on cuDNN fixing that bug.